### DISCUSSION TITLE
Callback with essential code

### DISCUSSION THREAD

**Discussion by SamKG** [Lightning-AI/pytorch-lightning#6830] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
Hi,
I was wondering if there was a callback-like api for essential code? (i.e. code that modifies batches, or changes model behavior)

Currently we are doing this by explicitly keeping a list of callbacks in the LightningModule, but this is not very maintainable in the long run.

**Comment by justusschock** [comment]
Hi @SamKG . Actually there isn't. There is the plugin API, but if you use these, your changes won't work for all the training types that are currently implemented as plugins.

Usually this is done by hooks. What is your usecase for that?

**Reply by SamKG** [reply]
Thanks for the response!

We are trying to allow users to easily compose multiple techniques together. 

For example, we currently allow users to transform the batch of images input into the model using a list of callbacks in `training_step`. The way a user currently adds a transform to this step is to append it as a callback to a list called `_batch_transformations` on the LightningModule. 

Essentially, we are looking for a way to allow for multiple hooks to be added to a LightningModule after construction.
